group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,384 @@
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()));
+                } else {
+                    log.debug("[GroupId {}] Committing offsets {} for partition {}-{} from member {} with leader epoch {}.",
+                        request.groupId(), partition.committedOffset(), topic.name(), partition.partitionIndex(),
+                        request.memberId(), partition.committedLeaderEpoch());
+
+                    topicResponse.partitions().add(new OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OffsetAndMetadata offsetAndMetadata = OffsetAndMetadata.fromRequest(
+                        partition,
+                        currentTimeMs,
+                        expireTimestampMs
+                    );
+
+                    records.add(RecordHelpers.newOffsetCommitRecord(
+                        request.groupId(),
+                        topic.name(),
+                        partition.partitionIndex(),
+                        offsetAndMetadata,
+                        metadataImage.features().metadataVersion()
+                    ));
+                }
+            });
+        });
+
+        return new CoordinatorResult<>(records, response);
+    }
+
+    /**
+     * Replays OffsetCommitKey/Value to update or delete the corresponding offsets.
+     *
+     * @param key   A OffsetCommitKey key.
+     * @param value A OffsetCommitValue value.
+     */
+    public void replay(
+        OffsetCommitKey key,
+        OffsetCommitValue value
+    ) {
+        final String groupId = key.group();
+        final TopicPartition tp = new TopicPartition(key.topic(), key.partition());
+
+        if (value != null) {
+            // The generic or consumer group should exist when offsets are committed or
+            // replayed. However, it won't if the consumer commits offsets but does not
+            // use the membership functionality. In this case, we automatically create
+            // a so-called "simple consumer group". This is an empty generic group
+            // without a protocol type.
+            try {
+                groupMetadataManager.group(groupId);
+            } catch (GroupIdNotFoundException ex) {
+                groupMetadataManager.getOrMaybeCreateGenericGroup(groupId, true);
+            }
+
+            final OffsetAndMetadata offsetAndMetadata = OffsetAndMetadata.fromRecord(value);
+            TimelineHashMap<TopicPartition, OffsetAndMetadata> offsets = offsetsByGroup.get(groupId);
+            if (offsets == null) {
+                offsets = new TimelineHashMap<>(snapshotRegistry, 0);
+                offsetsByGroup.put(groupId, offsets);
+            }
+
+            offsets.put(tp, offsetAndMetadata);
```

Comment:
i just noticed that out of the timeline data structures used in the new group coordinator only `GroupMetadataManager#groups` is updated before committing the generated records. 

how come the usage here indicates that we're updating the last committed offset vs. in getOrCreateGenericGroup() which is updating to the latest offset which is not yet committed?

i think i might be getting confused on the timeline data structures 😅

Reply:
I might also be confused, but I thought all the timeline data structures are updated before committing. (And then we can roll back if the commit fails, etc)

Reply:
> i just noticed that out of the timeline data structures used in the new group coordinator only GroupMetadataManager#groups is updated before committing the generated records.

We always update before committing. This is how the timeline data structures are used. As I explained in a previous comment. 1) generate records + response; 2) replay; 3) write; 4) put response to purgatory; 5) wait on hwm advancement to complete the response.

Reply:
sorry for the confusion. that makes sense, i for some reason assumed it to be 1) -> 3) -> 2) -> 4).